contracts/account/paymaster/PaymasterERC20Guarantor.sol

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {ERC4337Utils, PackedUserOperation} from "@openzeppelin/contracts/account/utils/draft-ERC4337Utils.sol";
+import {IERC20, SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {PaymasterERC20} from "./PaymasterERC20.sol";
+
+/// @title PaymasterERC20Guarantor
+/// @notice A paymaster that allows users to guarantee their user operations.
+abstract contract PaymasterERC20Guarantor is PaymasterERC20 {
+    using SafeERC20 for IERC20;
+
+    event UserOperationGuaranteed(bytes32 indexed userOpHash, address indexed user, address indexed guarantor);
+
+    function _validatePaymasterUserOp(
+        PackedUserOperation calldata userOp,
+        bytes32 userOpHash,
+        uint256 maxCost
+    ) internal virtual override returns (bytes memory context, uint256 validationData) {
+        // emit additional `UserOperationGuaranteed` event in case there is a guarantor, default to inherited behavior otherwise.
+        (uint256 validationData_, address guarantor) = _fetchGuarantor(userOp);
+        if (validationData_ == ERC4337Utils.SIG_VALIDATION_SUCCESS && guarantor != address(0)) {
+            emit UserOperationGuaranteed(userOpHash, userOp.sender, guarantor);
+        }
+        return super._validatePaymasterUserOp(userOp, userOpHash, maxCost);
```

Comment:
If there is a guarantor setup, the prefund will still be paid by `_prefundPayer` ... That doesn't feel right. This is a huge footgun if devs forget to override one of them, or if there is any inconsistency in the override